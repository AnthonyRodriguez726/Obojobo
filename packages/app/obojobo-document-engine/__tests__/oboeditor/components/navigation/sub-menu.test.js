import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

import ClipboardUtil from '../../../../src/scripts/oboeditor/util/clipboard-util'
import Common from '../../../../src/scripts/common'
import EditorUtil from '../../../../src/scripts/oboeditor/util/editor-util'
import SubMenu from '../../../../src/scripts/oboeditor/components/navigation/sub-menu'
import MoreInfoBox from '../../../../src/scripts/oboeditor/components/navigation/more-info-box'

jest.mock('../../../../src/scripts/oboeditor/util/editor-util', () => ({
	renamePage: jest.fn(),
	movePage: jest.fn(),
	addPage: jest.fn(),
	rebuildMenu: jest.fn(),
	deletePage: jest.fn(),
}))

jest.mock('../../../../src/scripts/common', () => ({
	models: {
		OboModel: {}
	},
	components: {
		modal: {
			SimpleDialog: () => 'MockSimpleDialog'
		}
	},
	util: {
		ModalUtil: {
			hide: jest.fn(),
			show: jest.fn()
		}
	},
	Registry: {
		getItems: funct => {
			return funct({
				values: () => [
					{
						isInsertable: true,
						icon: null,
						name: 'mockItem',
						templateObject: 'mockNode',
						cloneBlankNode: () => ({ type: 'mockNode' })
					}
				]
			})
		}
	}
}))

jest.mock('../../../../src/scripts/oboeditor/util/editor-util')
jest.mock('../../../../src/scripts/oboeditor/stores/editor-store', () => ({
	state: { startingId: null }
}))
// jest.mock('src/scripts/oboeditor/util/clipboard-util')
// jest.mock('src/scripts/common/util/modal-util')
jest.mock('../../../../src/scripts/oboeditor/components/navigation/more-info-box', () => {
	return props => <div mockname="MockMoreInfoBox" >{JSON.stringify(props, null, 2)}</div>
})

jest.useFakeTimers()

describe('SubMenu', () => {
	beforeEach(() => {
		jest.clearAllMocks()

		let mockGet = key => {
			switch(key){
				case 'type':
					return 'mock-type'
				case 'content':
					return 'mock-content'
			}
		}

		Common.models.OboModel.models = {
			'5': {
				isFirst: () => true,
				isLast: () => false,
				getIndex: () => 0,
				get: mockGet
			},
			'6': {
				isFirst: () => false,
				isLast: () => true,
				getIndex: () => 1,
				get: mockGet
			},
			'7': {
				isFirst: () => false,
				isLast: () => false,
				getIndex: () => 2,
				get: mockGet
			}
		}
	})

	test('SubMenu component', () => {
		const itemList = [
			{
				id: 5,
				type: 'link',
				label: 'label5',
				contentType: 'Page',
				flags: {
					assessment: false
				}
			}
		]
		const component = renderer.create(<SubMenu index={0} isSelected={false} list={itemList} />)

		const tree = component.toJSON()
		expect(tree).toMatchSnapshot()
	})

	test('SubMenu component selected with no contentType', () => {
		const itemList = [
			{
				id: 6,
				type: 'link',
				label: 'label6',
				flags: {
					assessment: false
				}
			}
		]
		const component = renderer.create(<SubMenu index={0} isSelected={true} list={itemList} />)

		const tree = component.toJSON()
		expect(tree).toMatchSnapshot()
	})

	test('SubMenu component calls parent onClick function', () => {
		const itemList = [
			{
				id: 6,
				type: 'link',
				label: 'label6',
				flags: {
					assessment: false
				}
			}
		]
		const parentOnClick = jest.fn()
		const component = mount(
			<SubMenu index={0} isSelected={true} list={itemList} onClick={parentOnClick} />
		)

		component
			.find('button') // [Link, Move Up, Move Down, Edit Name, Set Start, Delete, ID]
			.at(0)
			.simulate('click')

		expect(parentOnClick).toHaveBeenCalled()
	})

	test.skip('SubMenu component moves page up', () => {
		const itemList = [
			{
				id: 7,
				type: 'link',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		]
		const parentOnClick = jest.fn()
		const component = mount(
			<SubMenu index={0} isSelected={true} list={itemList} onClick={parentOnClick} />
		)

		component
			.find('button') // [Link, Move Up, Move Down, Edit Name, Set Start, Delete, ID]
			.at(1)
			.simulate('click')

		expect(EditorUtil.movePage).toHaveBeenCalledWith(7, 1)
	})

	test.skip('SubMenu component moves page down', () => {
		const itemList = [
			{
				id: 7,
				type: 'link',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		]
		const parentOnClick = jest.fn()
		const component = mount(
			<SubMenu index={0} isSelected={true} list={itemList} onClick={parentOnClick} />
		)

		component
			.find('button') // [Link, Move Up, Move Down, Edit Name, Set Start, Delete, ID]
			.at(2)
			.simulate('click')

		expect(EditorUtil.movePage).toHaveBeenCalledWith(7, 3)
	})

	test.skip('SubMenu component edits page name', () => {
		const itemList = [
			{
				id: 7,
				type: 'link',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		]
		const parentOnClick = jest.fn()
		const component = mount(
			<SubMenu index={0} isSelected={true} list={itemList} onClick={parentOnClick} />
		)

		component
			.find('button') // [Link, Move Up, Move Down, Edit Name, Set Start, Delete, ID]
			.at(3)
			.simulate('click')

		expect(ModalUtil.show).toHaveBeenCalled()
	})

	test('renamePage edits page name with empty label', () => {
		const itemList = [
			{
				id: 7,
				type: 'link',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		]
		const parentOnClick = jest.fn()
		const component = mount(
			<SubMenu index={0} isSelected={true} list={itemList} onClick={parentOnClick} />
		)

		component.instance().renamePage(7, '  ')

		expect(EditorUtil.renamePage).toHaveBeenCalled()
	})

	test('renamePage edits page name', () => {
		const itemList = [
			{
				id: 7,
				type: 'link',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		]
		const parentOnClick = jest.fn()
		const component = mount(
			<SubMenu index={0} isSelected={true} list={itemList} onClick={parentOnClick} />
		)

		component.instance().renamePage(7, 'mock title')

		expect(EditorUtil.renamePage).toHaveBeenCalled()
	})

	test('SubMenu component deletes page', () => {
		const itemList = [
			{
				id: 7,
				type: 'link',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		]

		const component = mount(
			<SubMenu index={0} list={itemList} />
		)

		component.instance().deletePage()

		expect(EditorUtil.deletePage).toHaveBeenCalled()
	})

	test.skip('SubMenu component copies id to clipboard', () => {
		const itemList = [
			{
				id: 7,
				type: 'link',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		]
		const parentOnClick = jest.fn()
		const component = mount(
			<SubMenu index={0} isSelected={true} list={itemList} onClick={parentOnClick} />
		)

		component
			.find('button') // [Link, Move Up, Move Down, Edit Name, Set Start, Delete, ID]
			.at(6)
			.simulate('click')

		expect(ClipboardUtil.copyToClipboard).toHaveBeenCalled()
	})

	test.skip('SubMenu component sets start page', () => {
		const itemList = [
			{
				id: 7,
				type: 'link',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		]
		const parentOnClick = jest.fn()
		const component = mount(
			<SubMenu index={0} isSelected={true} list={itemList} onClick={parentOnClick} />
		)

		component
			.find('button') // [Link, Move Up, Move Down, Edit Name, Set Start, Delete, ID]
			.at(4)
			.simulate('click')

		expect(EditorUtil.setStartPage).toHaveBeenCalled()
	})

	test.skip('SubMenu component opens menu', () => {
		const itemList = [
			{
				id: 7,
				type: 'link',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		]
		const parentOnClick = jest.fn()
		const component = mount(
			<SubMenu index={0} isSelected={true} list={itemList} onClick={parentOnClick} />
		)

		const html = component
			.find('li')
			.at(0)
			.simulate('keyDown', {
				key: 'ArrowRight'
			})
			.html()

		expect(html).toMatchSnapshot()
	})

	test.skip('SubMenu component closes menu', () => {
		const itemList = [
			{
				id: 7,
				type: 'link',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		]
		const parentOnClick = jest.fn()
		const component = mount(
			<SubMenu index={0} isSelected={true} list={itemList} onClick={parentOnClick} />
		)

		const html = component
			.find('li')
			.at(0)
			.simulate('keyDown', {
				key: 'ArrowLeft'
			})
			.html()

		expect(html).toMatchSnapshot()
	})

	test.skip('SubMenu component moves down through the menu', () => {
		const itemList = [
			{
				id: 7,
				type: 'link',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		]
		const parentOnClick = jest.fn()
		const component = mount(
			<SubMenu index={0} isSelected={true} list={itemList} onClick={parentOnClick} />
		)

		const html = component
			.find('li')
			.at(0)
			.simulate('keyDown', {
				key: 'ArrowDown'
			})
			.html()

		expect(html).toMatchSnapshot()
	})

	test.skip('SubMenu component moves up through the menu', () => {
		const itemList = [
			{
				id: 7,
				type: 'link',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		]
		const parentOnClick = jest.fn()
		const component = mount(
			<SubMenu index={0} isSelected={true} list={itemList} onClick={parentOnClick} />
		)

		const html = component
			.find('li')
			.at(0)
			.simulate('keyDown', {
				key: 'ArrowUp'
			})
			.html()

		expect(html).toMatchSnapshot()
	})

	test.skip('SubMenu component closes menu when unfocused', () => {
		const itemList = [
			{
				id: 7,
				type: 'link',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		]
		const parentOnClick = jest.fn()
		const component = mount(
			<SubMenu index={0} isSelected={true} list={itemList} onClick={parentOnClick} />
		)

		const html = component
			.find('li')
			.at(0)
			.simulate('blur')
			.html()

		jest.runAllTimers()

		expect(html).toMatchSnapshot()
	})

	test.skip('SubMenu component cancels menu closure when focused', () => {
		const itemList = [
			{
				id: 7,
				type: 'link',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		]
		const parentOnClick = jest.fn()
		const component = mount(
			<SubMenu index={0} isSelected={true} list={itemList} onClick={parentOnClick} />
		)

		const html = component
			.find('li')
			.at(0)
			.simulate('focus')
			.html()

		expect(html).toMatchSnapshot()
	})
})
