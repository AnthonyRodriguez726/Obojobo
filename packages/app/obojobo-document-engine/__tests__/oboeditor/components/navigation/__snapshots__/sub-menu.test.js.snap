// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubMenu SubMenu component 1`] = `
<li
  className="editor--editor-nav--submenu link is-not-selected is-not-assessment is-first-in-list is-last-in-list"
>
  <button
    aria-label="Go to Page: label5"
  >
    <span>
      label5
    </span>
  </button>
</li>
`;

exports[`SubMenu SubMenu component cancels menu closure when focused 1`] = `"<li class=\\"editor--editor-nav--submenu link is-selected is-not-assessment is-first-in-list is-last-in-list\\"><button aria-label=\\"Currently on label7\\"><span>label7</span></button><div class=\\"more-info undefined\\"><button class=\\"more-info-button \\"><svg viewBox=\\"0 0 214 214\\" xmlns=\\"http://www.w3.org/2000/svg\\" stroke=\\"#000\\" fill=\\"none\\"><g><g id=\\"svg_8\\"><ellipse ry=\\"100\\" rx=\\"100\\" cy=\\"107\\" cx=\\"107\\" stroke-width=\\"14\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"57\\" stroke-width=\\"16\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"107\\" stroke-width=\\"16\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"157\\" stroke-width=\\"16\\"></ellipse></g></g></svg></button></div><div class=\\"add-page\\"><button>+ Page</button></div></li>"`;

exports[`SubMenu SubMenu component closes menu 1`] = `"<li class=\\"editor--editor-nav--submenu link is-selected is-not-assessment is-first-in-list is-last-in-list\\"><button aria-label=\\"Currently on label7\\"><span>label7</span></button><div class=\\"more-info undefined\\"><button class=\\"more-info-button \\"><svg viewBox=\\"0 0 214 214\\" xmlns=\\"http://www.w3.org/2000/svg\\" stroke=\\"#000\\" fill=\\"none\\"><g><g id=\\"svg_8\\"><ellipse ry=\\"100\\" rx=\\"100\\" cy=\\"107\\" cx=\\"107\\" stroke-width=\\"14\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"57\\" stroke-width=\\"16\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"107\\" stroke-width=\\"16\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"157\\" stroke-width=\\"16\\"></ellipse></g></g></svg></button></div><div class=\\"add-page\\"><button>+ Page</button></div></li>"`;

exports[`SubMenu SubMenu component closes menu when unfocused 1`] = `"<li class=\\"editor--editor-nav--submenu link is-selected is-not-assessment is-first-in-list is-last-in-list\\"><button aria-label=\\"Currently on label7\\"><span>label7</span></button><div class=\\"more-info undefined\\"><button class=\\"more-info-button \\"><svg viewBox=\\"0 0 214 214\\" xmlns=\\"http://www.w3.org/2000/svg\\" stroke=\\"#000\\" fill=\\"none\\"><g><g id=\\"svg_8\\"><ellipse ry=\\"100\\" rx=\\"100\\" cy=\\"107\\" cx=\\"107\\" stroke-width=\\"14\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"57\\" stroke-width=\\"16\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"107\\" stroke-width=\\"16\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"157\\" stroke-width=\\"16\\"></ellipse></g></g></svg></button></div><div class=\\"add-page\\"><button>+ Page</button></div></li>"`;

exports[`SubMenu SubMenu component moves down through the menu 1`] = `"<li class=\\"editor--editor-nav--submenu link is-selected is-not-assessment is-first-in-list is-last-in-list\\"><button aria-label=\\"Currently on label7\\"><span>label7</span></button><div class=\\"more-info undefined\\"><button class=\\"more-info-button \\"><svg viewBox=\\"0 0 214 214\\" xmlns=\\"http://www.w3.org/2000/svg\\" stroke=\\"#000\\" fill=\\"none\\"><g><g id=\\"svg_8\\"><ellipse ry=\\"100\\" rx=\\"100\\" cy=\\"107\\" cx=\\"107\\" stroke-width=\\"14\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"57\\" stroke-width=\\"16\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"107\\" stroke-width=\\"16\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"157\\" stroke-width=\\"16\\"></ellipse></g></g></svg></button></div><div class=\\"add-page\\"><button>+ Page</button></div></li>"`;

exports[`SubMenu SubMenu component moves up through the menu 1`] = `"<li class=\\"editor--editor-nav--submenu link is-selected is-not-assessment is-first-in-list is-last-in-list\\"><button aria-label=\\"Currently on label7\\"><span>label7</span></button><div class=\\"more-info undefined\\"><button class=\\"more-info-button \\"><svg viewBox=\\"0 0 214 214\\" xmlns=\\"http://www.w3.org/2000/svg\\" stroke=\\"#000\\" fill=\\"none\\"><g><g id=\\"svg_8\\"><ellipse ry=\\"100\\" rx=\\"100\\" cy=\\"107\\" cx=\\"107\\" stroke-width=\\"14\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"57\\" stroke-width=\\"16\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"107\\" stroke-width=\\"16\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"157\\" stroke-width=\\"16\\"></ellipse></g></g></svg></button></div><div class=\\"add-page\\"><button>+ Page</button></div></li>"`;

exports[`SubMenu SubMenu component opens menu 1`] = `"<li class=\\"editor--editor-nav--submenu link is-selected is-not-assessment is-first-in-list is-last-in-list\\"><button aria-label=\\"Currently on label7\\"><span>label7</span></button><div class=\\"more-info undefined\\"><button class=\\"more-info-button \\"><svg viewBox=\\"0 0 214 214\\" xmlns=\\"http://www.w3.org/2000/svg\\" stroke=\\"#000\\" fill=\\"none\\"><g><g id=\\"svg_8\\"><ellipse ry=\\"100\\" rx=\\"100\\" cy=\\"107\\" cx=\\"107\\" stroke-width=\\"14\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"57\\" stroke-width=\\"16\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"107\\" stroke-width=\\"16\\"></ellipse><ellipse ry=\\"7\\" rx=\\"7\\" cy=\\"107\\" cx=\\"157\\" stroke-width=\\"16\\"></ellipse></g></g></svg></button></div><div class=\\"add-page\\"><button>+ Page</button></div></li>"`;

exports[`SubMenu SubMenu component selected with no contentType 1`] = `
<li
  className="editor--editor-nav--submenu link is-selected is-not-assessment is-first-in-list is-last-in-list"
>
  <button
    aria-label="Currently on label6"
  >
    <span>
      label6
    </span>
  </button>
  <div
    mockname="MockMoreInfoBox"
  >
    {
  "id": 6,
  "index": 1,
  "isFirst": false,
  "isLast": true,
  "type": "mock-type",
  "content": "mock-content",
  "contentDescription": [
    {
      "name": "title",
      "description": "Title",
      "type": "input"
    }
  ],
  "showMoveButtons": true
}
  </div>
  <div
    className="add-page"
  >
    <button
      onClick={[Function]}
    >
      + Page
    </button>
  </div>
</li>
`;
